core/finality-grandpa/src/communication/gossip.rs

```diff
@@ -512,7 +546,58 @@ impl<Block: BlockT> Inner<Block> {
 		self.multicast_neighbor_packet()
 	}
 
-	fn consider_vote(&self, round: Round, set_id: SetId) -> Consider {
+	fn consider_vote(&mut self, who: &PeerId, round: Round, set_id: SetId, msg: &SignedMessage<Block>) -> Consider {
+		// A tally of messages received per peer, per voter, per round.
+		let peer_tally_for_round = match self.incoming_msg_tally.get_mut(&(round, set_id)) {
+			Some(tally_for_round) =>  tally_for_round,
+			None => {
+				// We don't know about this round,
+				// let the local-view handle it and likely treat it as past or future.
+				return self.local_view.as_ref().map(|v| v.consider_vote(round, set_id))
+					.unwrap_or(Consider::RejectOutOfScope)
+			}
+		};
+		let mut per_peer_tally = peer_tally_for_round.entry((msg.id.clone(), who.clone())).or_insert(Default::default());
+
+		// A tally of messages we sent out, per peer, per voter, per round
+		let outgoing_tally_for_round = self.outgoing_msg_tally
+			.get_mut(&(round, set_id))
+			.expect("If incoming votes knows about this round, so should outgoing ones; qed");
+		let outgoing_tally = outgoing_tally_for_round.entry((msg.id.clone(), who.clone())).or_insert(Default::default());
+
+		match &msg.message {
+			PrimaryPropose(_propose) => {
+				if per_peer_tally.primary_proposals < 3 {
+					per_peer_tally.primary_proposals += 1;
+				}
+			},
+			Prevote(_prevote) => {
+				if per_peer_tally.pre_votes < 3 {
+					per_peer_tally.pre_votes += 1;
+				}
+			},
+			Precommit(_precommit) => {
+				if per_peer_tally.pre_commits < 3 {
+					per_peer_tally.pre_commits += 1;
+				}
+			},
+		}
+
+		// We reject all messages for a given voter,
+		// once we're received more than two messages of any kind for that voter.
+		//
+		// We report peers who send us a third or more of message of any kind,
+		// when we've already exchanged two messages of any kind for that voter to that peer.
+		let should_report = (outgoing_tally.primary_proposals + per_peer_tally.primary_proposals) > 2 ||
+				(outgoing_tally.pre_votes + per_peer_tally.pre_votes) > 2 ||
```

Comment:
```suggestion
			(outgoing_tally.pre_votes + per_peer_tally.pre_votes) > 2 ||
```

Reply:
Are we sure the peer won't send us the same message we sent him?

Reply:
If the peer is running this code, there is a check at `message_allowed` that is essentially the mirror or this one, so the idea is that we won't send out what we wouldn't accept, based on what we've received and sent out so far. 

So I think so, yes. Do you have something else in mind?